README.md
# stactools-noaa-mrms-qpe

[![PyPI](https://img.shields.io/pypi/v/stactools-noaa-mrms-qpe)](https://pypi.org/project/stactools-noaa-mrms-qpe/)

- Name: noaa-mrms-qpe
- Package: `stactools.noaa_mrms_qpe`
- PyPI: <https://pypi.org/project/stactools-noaa-mrms-qpe/>
- Owner: @m-mohr
- Dataset homepage: <https://mrms.nssl.noaa.gov>
- STAC extensions used:
  - [classification](https://github.com/stac-extensions/classification/)
  - [NOAA MRMS QPE](https://github.com/stac-extensions/noaa-mrms-qpe)
  - [raster](https://github.com/stac-extensions/raster/)
  - [proj](https://github.com/stac-extensions/projection/)
- Extra fields:
  - See the [NOAA MRMS QPE extension](https://github.com/stac-extensions/noaa-mrms-qpe) for details

A stactools package for NOAA's Multi-Radar Multi-Sensor (MRMS) Quantitative Precipitation Estimation (QPE) dataset.

This package can generate STAC files from (gzipped) GRIB2 files that link to the original GRIB2 files and/or
to Cloud-Optimized GeoTiffs (COGs) in the original or any other EPSG projection.

## STAC Examples

- [Collection](examples/collection.json)
- [Item for continental US](examples/item-conus.json)
- [Item for GUAM](examples/item-guam.json)
- [Browse the example in a human-readable form](https://radiantearth.github.io/stac-browser/#/external/raw.githubusercontent.com/stactools-packages/noaa-mrms-qpe/main/examples/collection.json)

## Installation
```shell
pip install stactools-noaa-mrms-qpe
```

## Command-line Usage

### Collection

Create a collection, e.g. 24-hour Pass 2:

```shell
stac noaa-mrms-qpe create-collection collection.json --period 24 --pass_no 2
```

Get information about all options for collection creation:

```shell
stac noaa-mrms-qpe create-collection --help
```

### Item

Create an item for continentel US with a GRIB2 and COG asset:

```shell
stac noaa-mrms-qpe create-item MRMS_MultiSensor_QPE_24H_Pass2_00.00_20220530-120000.grib2.gz item_grib.json --collection collection.json
```

Create an item for ALASKA with only a COG asset converted to EPSG:3857:

```shell
stac noaa-mrms-qpe create-item MRMS_MultiSensor_QPE_24H_Pass2_00.00_20220530-120000.grib2.gz item.json --aoi ALASKA --collection collection.json --nogrib TRUE --epsg 3857
```

Get information about all options for item creation:

```shell
stac noaa-mrms-qpe create-item --help
```

Use `stac noaa-mrms-qpe --help` to see all subcommands and options.

*Note: This package can only read files that contain the timestamp in the file name. It can NOT read the files that contain `latest` instead of a timestamp in the file name.*

## Contributing

We use [pre-commit](https://pre-commit.com/) to check any changes.
To set up your development environment:

```shell
$ pip install -e .
$ pip install -r requirements-dev.txt
$ pre-commit install
```

To check all files:

```shell
$ pre-commit run --all-files
```

To run the tests:

```shell
$ pytest -vv
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## [0.3.1]

### Fixed

- Capture the minutes and seconds from the file names as they are not always 0

## [0.3.0]

### Added

- New property `noaa_mrms_qpe:region`

### Changed

- Moved extension to separate repository: <https://github.com/stac-extensions/noaa-mrms-qpe>
- Always add `proj:epsg` to Item to fulfill a proj extension "requirement" (schema issue in 1.0.0).

### Fixed

- Changed casing of Cloud-Optimized in the asset titles

## [0.2.0]

### Added

- New property `noaa_mrms_qpe:region` added to Items and Collections
- Officially add support for Python 3.10

### Fixed

- remove unused portions of CI that are causing problems

## [0.1.1]

### Changed

- cast numpy float64s to float

### Fixed

- remove unused portions of CI that are causing problems

## [0.1.0]

- First release

[Unreleased]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/main/>
[0.3.1]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/v0.3.1/>
[0.3.0]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/v0.3.0/>
[0.2.0]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/v0.2.0/>
[0.1.1]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/v0.1.1/>
[0.1.0]: <https://github.com/stactools-packages/noaa-mrms-qpe/tree/v0.1.0/>


tests/test_stac.py
import os.path
import shutil
import unittest
from datetime import datetime, timezone
from tempfile import TemporaryDirectory
from typing import Any, Dict, List, Optional

from pystac import Collection, Item

from stactools.noaa_mrms_qpe import constants, stac

PERIODS: List[int] = [1, 3, 6, 12, 24, 48, 72]
PASS_NUMBERS: List[int] = [1, 2]

COG_MEDIATYPE: str = "image/tiff; application=geotiff; profile=cloud-optimized"
GRIB_MEDIATYPE: str = "application/wmo-GRIB2"

TEST_COLLECTIONS: List[Dict[str, Any]] = []
for period in PERIODS:
    for pass_no in PASS_NUMBERS:
        for nocog in [True, False]:
            for nogrib in [True, False]:
                TEST_COLLECTIONS.append(
                    {
                        "period": period,
                        "pass_no": pass_no,
                        "nocog": nocog,
                        "nogrib": nogrib,
                    }
                )

TEST_FILES: List[List[Any]] = [
    [
        "ALASKA",  # aoi
        "MRMS_MultiSensor_QPE_12H_Pass2_00.00_20220602-000000",  # filename
        12,  # period
        2,  # pass no
        [2022, 6, 2, 0],  # datetime
        True,  # gzip
        True,  # nocog
        True,  # nogrib
        None,  # collection path
        False,  # has nodata values
    ],
    [
        "CARIB",
        "MRMS_MultiSensor_QPE_24H_Pass2_00.00_20220602-030000",
        24,
        2,
        [2022, 6, 2, 3],
        True,
        3857,
        False,
        None,
        True,
    ],
    [
        "CONUS",
        "MRMS_MultiSensor_QPE_01H_Pass2_00.00_20220602-120000",
        1,
        2,
        [2022, 6, 2, 12],
        True,
        False,
        False,
        None,
        False,
    ],
    [
        "GUAM",
        "MRMS_MultiSensor_QPE_01H_Pass1_00.00_20220601-120000",
        1,
        1,
        [2022, 6, 1, 12],
        True,
        False,
        False,
        "./tests/data-files/collection-1-1.json",
        True,
    ],
    [
        "HAWAII",
        "MRMS_MultiSensor_QPE_72H_Pass2_00.00_20220601-230000",
        72,
        2,
        [2022, 6, 1, 23],
        False,
        True,
        False,
        None,
        False,
    ],
]


class StacTest(unittest.TestCase):
    def test_create_collection(self) -> None:
        for test_data in TEST_COLLECTIONS:
            with self.subTest(test_data=test_data):
                period: int = test_data["period"]
                pass_no: int = test_data["pass_no"]
                nocog: bool = test_data["nocog"]
                nogrib: bool = test_data["nogrib"]

                # untested right now: parameters start_time, thumbnail
                collection = stac.create_collection(
                    period=period, pass_no=pass_no, nocog=nocog, nogrib=nogrib
                )
                collection.set_self_href("")

                collection.validate()

                collection_dict = collection.to_dict()

                self.assertEqual(
                    collection.summaries.get_list("noaa_mrms_qpe:pass"), [pass_no]
                )
                self.assertEqual(
                    collection.summaries.get_list("noaa_mrms_qpe:period"), [period]
                )

                self.assertTrue("item_assets" in collection_dict)
                assets: Dict[str, Dict[str, Any]] = collection_dict["item_assets"]

                asset_count = 2
                if nocog:
                    asset_count -= 1
                if nogrib:
                    asset_count -= 1
                self.assertEqual(len(assets), asset_count)

                # Check COG asset
                if nocog:
                    self.assertFalse("cog" in assets)
                else:
                    self.assertTrue("cog" in assets)
                    cog_asset = assets["cog"]
                    self.assertEqual(cog_asset["type"], COG_MEDIATYPE)
                    self.assertTrue("raster:bands" in cog_asset)
                    self.assertEqual(len(cog_asset["raster:bands"]), 1)
                    cog_band = cog_asset["raster:bands"][0]
                    self.assertEqual(cog_band["spatial_resolution"], 1000)
                    self.assertEqual(cog_band["unit"], "mm")
                    self.assertEqual(cog_band["data_type"], "float64")
                    self.assertFalse("classification:classes" in cog_band)
                    self.assertFalse("nodata" in cog_band)

                # Check GRIB2 asset
                if nogrib:
                    self.assertFalse("grib2" in assets)
                else:
                    self.assertTrue("grib2" in assets)
                    grib_asset = assets["grib2"]
                    self.assertEqual(grib_asset["type"], GRIB_MEDIATYPE)
                    self.assertTrue("raster:bands" in grib_asset)
                    self.assertEqual(len(grib_asset["raster:bands"]), 1)
                    grib_band = grib_asset["raster:bands"][0]
                    self.assertEqual(grib_band["spatial_resolution"], 1000)
                    self.assertEqual(grib_band["unit"], "mm")
                    self.assertEqual(grib_band["data_type"], "float64")
                    self.assertFalse("classification:classes" in grib_band)
                    self.assertFalse("nodata" in grib_band)

    def test_create_item(self) -> None:
        for test_data in TEST_FILES:
            with self.subTest(test_data=test_data):
                folder: str = test_data[0]
                id: str = test_data[1]
                period: int = test_data[2]
                pass_no: int = test_data[3]
                dt: List[int] = test_data[4]
                gzip: bool = test_data[5]
                nocog: bool = False
                epsg: int = 0
                if isinstance(test_data[6], bool):
                    nocog = test_data[6]
                else:
                    epsg = test_data[6]
                nogrib: bool = test_data[7]
                collection_path: Optional[str] = test_data[8]
                nodata: bool = test_data[9]
                proj_toplevel: bool = epsg == 0 or nogrib or nocog

                src_dir = f"./tests/data-files/{folder}"
                filename = f"{id}.grib2"
                if gzip:
                    filename += ".gz"

                item: Optional[Item] = None
                with TemporaryDirectory() as tmp_dir:
                    src_data_file = os.path.join(src_dir, filename)
                    dest_data_file = os.path.join(tmp_dir, filename)
                    shutil.copyfile(src_data_file, dest_data_file)

                    collection = None
                    if collection_path is not None:
                        collection = Collection.from_file(collection_path)

                    item = stac.create_item(
                        dest_data_file,
                        aoi=constants.AOI[folder],
                        nogrib=nogrib,
                        nocog=nocog,
                        collection=collection,
                        epsg=epsg,
                    )

                item.validate()

                self.assertIsNotNone(item)
                self.assertEqual(item.id, f"{folder}_{id}")
                self.assertEqual(item.properties["noaa_mrms_qpe:pass"], pass_no)
                self.assertEqual(item.properties["noaa_mrms_qpe:period"], period)
                self.assertEqual(
                    item.properties["noaa_mrms_qpe:region"], constants.AOI[folder]
                )
                self.assertTrue("description" in item.properties)
                ref_dt = datetime(
                    dt[0], dt[1], dt[2], dt[3], 0, 0, 0, tzinfo=timezone.utc
                )
                self.assertEqual(item.datetime, ref_dt)

                self.assertEqual("proj:projjson" in item.properties, proj_toplevel)
                self.assertIsNone(item.properties["proj:epsg"])
                if proj_toplevel:
                    self.assertIsInstance(item.properties["proj:projjson"], dict)

                if collection is not None:
                    self.assertEqual(item.collection_id, collection.id)
                else:
                    self.assertIsNone(item.collection_id)

                asset_count = 2
                if nocog:
                    asset_count -= 1
                if nogrib:
                    asset_count -= 1
                self.assertEqual(len(item.assets), asset_count)

                # Check COG asset
                if nocog:
                    self.assertFalse("cog" in item.assets)
                else:
                    cog_asset = item.assets["cog"].to_dict()
                    self.assertEqual(cog_asset["type"], COG_MEDIATYPE)
                    self.assertTrue("title" in cog_asset)
                    self.assertEqual(len(cog_asset["roles"]), 2)
                    self.assertEqual(len(cog_asset["proj:shape"]), 2)
                    self.assertEqual(len(cog_asset["proj:transform"]), 6)
                    self.assertEqual("proj:epsg" in cog_asset, not proj_toplevel)
                    if not proj_toplevel:
                        if epsg > 0:
                            self.assertEqual(cog_asset["proj:epsg"], epsg)
                        else:
                            self.assertIsNone(cog_asset["proj:epsg"])
                            self.assertIsInstance(cog_asset["proj:projjson"], dict)

                    self.assertTrue("raster:bands" in cog_asset)
                    self.assertEqual(len(cog_asset["raster:bands"]), 1)
                    cog_band = cog_asset["raster:bands"][0]
                    self.assertEqual(cog_band["spatial_resolution"], 1000)
                    self.assertEqual(cog_band["unit"], "mm")
                    self.assertEqual(cog_band["data_type"], "float64")

                    self.assertTrue("statistics" in cog_band)
                    self.assertEqual(cog_band["statistics"]["minimum"], 0)
                    self.assertGreaterEqual(cog_band["statistics"]["maximum"], 0)

                    self.assertEqual("classification:classes" in cog_band, nodata)
                    if nodata:
                        self.assertEqual(len(cog_band["classification:classes"]), 1)
                        class1 = cog_band["classification:classes"][0]
                        self.assertEqual(class1["value"], -1.0)
                        self.assertTrue("name" in class1)
                        self.assertTrue("description" in class1)
                        self.assertTrue(class1["nodata"])

                        self.assertEqual(cog_band["nodata"], -1.0)

                # Check GRIB2 asset
                if nogrib:
                    self.assertFalse("grib2" in item.assets)
                else:
                    grib_asset = item.assets["grib2"].to_dict()
                    self.assertEqual(grib_asset["type"], GRIB_MEDIATYPE)
                    self.assertTrue("title" in grib_asset)
                    self.assertEqual(len(grib_asset["roles"]), 2)
                    self.assertEqual(len(grib_asset["proj:shape"]), 2)
                    self.assertEqual(len(grib_asset["proj:transform"]), 6)
                    self.assertEqual("proj:epsg" in grib_asset, not proj_toplevel)
                    self.assertEqual("proj:projjson" in grib_asset, not proj_toplevel)
                    if not proj_toplevel:
                        self.assertIsNone(grib_asset["proj:epsg"])
                        self.assertIsInstance(grib_asset["proj:projjson"], dict)

                    self.assertTrue("raster:bands" in grib_asset)
                    self.assertEqual(len(grib_asset["raster:bands"]), 1)
                    grib_band = grib_asset["raster:bands"][0]
                    self.assertEqual(grib_band["spatial_resolution"], 1000)
                    self.assertEqual(grib_band["unit"], "mm")
                    self.assertEqual(grib_band["data_type"], "float64")

                    self.assertTrue("statistics" in grib_band)
                    self.assertEqual(grib_band["statistics"]["minimum"], 0)
                    self.assertGreaterEqual(grib_band["statistics"]["maximum"], 0)

                    self.assertEqual("classification:classes" in grib_band, nodata)
                    if nodata:
                        self.assertGreaterEqual(
                            len(grib_band["classification:classes"]), 1
                        )
                        class1 = grib_band["classification:classes"][0]
                        self.assertIn(class1["value"], [-1.0, -3.0, -999.0])
                        self.assertTrue("name" in class1)
                        self.assertTrue("description" in class1)
                        self.assertTrue(class1["nodata"])

                        if len(grib_band["classification:classes"]) == 1:
                            self.assertTrue("nodata" in grib_band)

    def test_datetime(self) -> None:
        src_dir = "./tests/data-files/ALASKA"
        filename = "MRMS_MultiSensor_QPE_01H_Pass1_00.00_20221024-015800.grib2.gz"

        with TemporaryDirectory() as tmp_dir:
            src_data_file = os.path.join(src_dir, filename)
            dest_data_file = os.path.join(tmp_dir, filename)
            shutil.copyfile(src_data_file, dest_data_file)

            item = stac.create_item(
                dest_data_file,
                aoi=constants.AOI["ALASKA"],
                nogrib=True,
                nocog=True,
            )

        item.validate()
        if item.datetime is not None:
            item_datetime = item.datetime
        self.assertEqual(item_datetime.year, 2022)
        self.assertEqual(item_datetime.month, 10)
        self.assertEqual(item_datetime.day, 24)
        self.assertEqual(item_datetime.hour, 1)
        self.assertEqual(item_datetime.minute, 58)
        self.assertEqual(item_datetime.second, 0)


tests/test_module.py
import unittest

import stactools.noaa_mrms_qpe


class TestModule(unittest.TestCase):
    def test_version(self) -> None:
        self.assertIsNotNone(stactools.noaa_mrms_qpe.__version__)


tests/test_commands.py
import json
import os.path
import shutil
from tempfile import TemporaryDirectory
from typing import Callable, List

from click import Command, Group
from deepdiff import DeepDiff
from stactools.testing.cli_test import CliTestCase

from stactools.noaa_mrms_qpe.commands import create_noaa_mrms_qpe_command


class CommandsTest(CliTestCase):
    def create_subcommand_functions(self) -> List[Callable[[Group], Command]]:
        return [create_noaa_mrms_qpe_command]

    def test_create_collection(self) -> None:
        with TemporaryDirectory() as tmp_dir:
            src_folder = "./tests/data-files/"
            src_file = os.path.join(src_folder, "collection-1-1.json")
            destination = os.path.join(tmp_dir, "collection-1-1.json")
            period = 1
            pass_no = 1

            result = self.run_command(
                f"noaa-mrms-qpe create-collection {destination} "
                f"--period {period} --pass_no {pass_no} --start_time 2022-01-01T00:00:00Z"
            )

            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(tmp_dir) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            collection = {}
            truth_collection = {}
            with open(destination) as f:
                collection = json.load(f)
            with open(src_file) as f:
                truth_collection = json.load(f)

            self.assertEqual(collection["id"], f"noaa-mrms-qpe-{period}h-pass{pass_no}")

            diff = DeepDiff(
                collection,
                truth_collection,
                ignore_order=True,
                exclude_regex_paths=r"root\['links'\]\[\d+\]\['href'\]",
            )
            self.assertEqual(diff, {})

    def test_create_item(self) -> None:
        with TemporaryDirectory() as tmp_dir:
            id = "MRMS_MultiSensor_QPE_01H_Pass1_00.00_20220601-120000"
            aoi = "GUAM"
            src_folder = f"./tests/data-files/{aoi}"
            src_data_filename = f"{id}.grib2.gz"
            stac_filename = f"{id}.json"

            src_data_file = os.path.join(src_folder, src_data_filename)
            dest_data_file = os.path.join(tmp_dir, src_data_filename)
            shutil.copyfile(src_data_file, dest_data_file)

            src_stac = os.path.join(src_folder, stac_filename)
            dest_stac = os.path.join(tmp_dir, stac_filename)

            result = self.run_command(
                f"noaa-mrms-qpe create-item {dest_data_file} {dest_stac} --aoi {aoi}"
            )
            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            files = os.listdir(tmp_dir)
            jsons = [p for p in files if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            item = {}
            truth_item = {}
            with open(dest_stac) as f:
                item = json.load(f)
            with open(src_stac) as f:
                truth_item = json.load(f)

            self.assertEqual(item["id"], f"{aoi}_{id}")

            diff = DeepDiff(
                item,
                truth_item,
                ignore_order=True,
                exclude_paths=[
                    "root['assets']['cog']['href']",
                    "root['assets']['grib2']['href']",
                ],
            )
            self.assertEqual(diff, {})


src/stactools/noaa_mrms_qpe/stac.py
import logging
import os
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Union

import numpy as np
import rasterio
from dateutil.parser import isoparse
from pystac import (
    Asset,
    CatalogType,
    Collection,
    Extent,
    Item,
    MediaType,
    SpatialExtent,
    Summaries,
    TemporalExtent,
)
from pystac.extensions.item_assets import AssetDefinition, ItemAssetsExtension
from pystac.extensions.projection import ProjectionExtension
from pystac.extensions.raster import DataType

from . import cog, constants
from .fileinfo import FileInfo

logger = logging.getLogger(__name__)


def create_collection(
    period: int,
    pass_no: int,
    thumbnail: str = "",
    nocog: bool = False,
    nogrib: bool = False,
    start_time: Optional[str] = None,
) -> Collection:
    """Create a STAC Collection for NOAA MRMS QPE sub-products.

    Args:
        period (int): The time period the sub-product is for (either 1, 3, 6, 12, 24, 48, or 72)
        pass_no (int): The pass number of the sub-product (either 1 or 2)
        thumbnail (str): URL for the PNG or JPEG collection thumbnail asset (none if empty)
        nocog (bool): If set to True, the collections does not include the COG-related metadata
        nogrib (bool): If set to True, the collections does not include the GRIB2-related metadata
        start_time (str): The start timestamp for the temporal extent, default to now.
            Timestamps consist of a date and time in UTC and must follow RFC 3339, section 5.6.

    Returns:
        Collection: STAC Collection object
    """
    # Time must be in UTC
    if start_time is None:
        start_datetime = datetime.now(tz=timezone.utc)
    else:
        start_datetime = isoparse(start_time)

    spatial_extents = list(constants.EXTENTS.values())
    extent = Extent(
        SpatialExtent(spatial_extents),
        TemporalExtent([[start_datetime, None]]),
    )

    keywords = [
        "NOAA",
        "MRMS",
        "QPE",
        "multi-radar",
        "multi-sensor",
        "precipitation",
        "{t}-hour".format(t=period),
    ]
    if not nogrib:
        keywords.append("GRIB2")
    if not nocog:
        keywords.append("COG")

    description = (
        "The Multi-Radar Multi-Sensor (MRMS) quantitative precipitation estimation "
        "(QPE) product is generated fully automatically from multiple sources to generate "
        "seamless, hourly 1 km mosaics over the US.\n\n"
        "**Note:** The data for Guam and the Caribbean Islands are [not multi-sensor products]"
        "(https://vlab.noaa.gov/documents/96675/666999/MS_DomainDiffernces.png) yet."
    )
    if pass_no == 1:
        description += (
            "\n\nThis is the {t}-hour pass 1 product with less latency (60 min), "
            "but less gauges (60-65 %)."
        )
    elif pass_no == 2:
        description += (
            "\n\nThis is the {t}-hour pass 2 product with more latency (120 min), "
            "but more gauges (99 %)."
        )

    summaries = Summaries({})
    summaries.add(constants.EXT_PASS, [pass_no])
    summaries.add(constants.EXT_PERIOD, [period])
    summaries.add(constants.EXT_REGION, [e.value for e in constants.AOI])

    collection = Collection(
        stac_extensions=[constants.EXTENSION],
        id="noaa-mrms-qpe-{t}h-pass{p}".format(t=period, p=pass_no),
        title="NOAA MRMS QPE {t}-hour Pass {p}".format(t=period, p=pass_no),
        description=description.format(t=period),
        keywords=keywords,
        license="proprietary",
        providers=constants.PROVIDERS,
        extent=extent,
        summaries=summaries,
        catalog_type=CatalogType.RELATIVE_PUBLISHED,
    )

    collection.add_link(constants.LINK_LICENSE)
    collection.add_link(constants.LINK_MRMS_HOME)
    collection.add_link(constants.LINK_MRMS_TECH_GUIDE)

    if len(thumbnail) > 0:
        if thumbnail.endswith(".png"):
            media_type = MediaType.PNG
        else:
            media_type = MediaType.JPEG

        collection.add_asset(
            "thumbnail",
            Asset(
                href=thumbnail,
                title="Preview",
                roles=["thumbnail"],
                media_type=media_type,
            ),
        )

    item_assets = {}

    # it seems the raster extension can't be added to an AssetDefintion
    # via RasterExtension.ext(data_asset, add_if_missing=True).
    # So RasterBand.create() etc. are not usable here
    collection.stac_extensions.append(constants.RASTER_EXTENSION_V11)

    def create_asset(media_type: str, roles: List[str], title: str) -> Dict[str, Any]:
        asset: Dict[str, Any] = {
            "roles": roles,
            "type": media_type,
            "raster:bands": [create_band()],
            "title": title,
        }
        return asset

    if not nocog:
        asset = create_asset(
            MediaType.COG, constants.COG_ROLES, constants.ASSET_COG_TITLE
        )
        item_assets[constants.ASSET_COG_KEY] = AssetDefinition(asset)

    if not nogrib:
        asset = create_asset(
            constants.GRIB2_MEDIATYPE,
            constants.GRIB2_ROLES,
            constants.ASSET_GRIB2_TITLE,
        )
        item_assets[constants.ASSET_GRIB2_KEY] = AssetDefinition(asset)

    item_assets_attrs = ItemAssetsExtension.ext(collection, add_if_missing=True)
    item_assets_attrs.item_assets = item_assets

    return collection


def create_item(
    asset_href: str,
    aoi: constants.AOI,
    collection: Optional[Collection] = None,
    nocog: bool = False,
    nogrib: bool = False,
    epsg: int = 0,
) -> Item:
    """Create a STAC Item

    This function should include logic to extract all relevant metadata from an
    asset, metadata asset, and/or a constants.py file.

    See `Item<https://pystac.readthedocs.io/en/latest/api.html#item>`_.

    Args:
        asset_href (str): The HREF pointing to an asset associated with the item
        aoi (AOI): The area of interest, either 'ALASKA', 'CONUS' (continental US),
            'CARIB' (Caribbean islands), 'GUAM' or 'HAWAII'
        collection (pystac.Collection): HREF to an existing collection
        nocog (bool): If set to True, no COG file is generated for the Item
        nogrib (bool): If set to True, the GRIB2 file is not added to the Item
        epsg (int): Converts the COG files to the given EPSG Code (e.g. 3857),
            doesn't reproject by default.

    Returns:
        Item: STAC Item object
    """

    basics = parse_filename(asset_href)
    id = aoi + "_" + basics.id

    bbox = constants.EXTENTS[aoi]

    description = "Multi-sensor accumulation {p}-hour ({t}-hour latency) [mm]".format(
        p=basics.period, t=basics.pass_no
    )

    properties = {
        constants.EXT_PASS: basics.pass_no,
        constants.EXT_PERIOD: basics.period,
        constants.EXT_REGION: aoi.value,
        "description": description,
    }

    item = Item(
        stac_extensions=[constants.EXTENSION],
        id=id,
        properties=properties,
        geometry=bbox_to_polygon(bbox),
        bbox=bbox,
        datetime=basics.datetime,
        collection=collection,
    )

    # Raster extension v1.1 not supported by PySTAC
    item.stac_extensions.append(constants.RASTER_EXTENSION_V11)
    # Classification extension v1.1 not supported by PySTAC
    item.stac_extensions.append(constants.CLASSIFICATION_EXTENSION_V11)

    # Projection extension for assets
    proj_attrs = ProjectionExtension.ext(item, add_if_missing=True)
    # Set CRS details globally if they are the same for COG and GRIB or only one of them is exposed.
    # Otherwise, we set the CRS information in the asset
    if epsg == 0 or nocog or nogrib:
        proj_attrs.epsg = None
        proj_attrs.projjson = constants.PROJJSON
    else:
        # Item validation fails if proj:epsg is not in the items due to a bug in the schema of the
        # projection extension, see https://github.com/stac-extensions/projection/issues/7
        # So the following line should be removed once the issue has been solved.
        proj_attrs.epsg = None

    def create_asset(
        href: str,
        media_type: str,
        roles: List[str],
        band: Dict[str, Any],
        crs: Union[Dict[str, Any], int],
        title: str,
    ) -> Asset:
        asset = Asset(href=href, media_type=media_type, roles=roles, title=title)

        isGRIB2 = media_type == constants.GRIB2_MEDIATYPE

        shape = None
        transform = None
        with rasterio.open(href) as dataset:
            if dataset.transform:
                transform = list(dataset.transform)[0:6]

            if len(dataset.shape) == 2:
                shape = [dataset.shape[1], dataset.shape[0]]

            data = dataset.read()
            valid_data = np.ma.masked_array(data, mask=(data < 0))  # type: ignore

            band["statistics"] = {
                "minimum": float(np.nanmin(valid_data)),
                "maximum": float(np.nanmax(valid_data)),
            }

            classes = []
            if isGRIB2:
                if np.any(data == -1.0):
                    classes.append(constants.GRIB2_CLASSIFICATION[0])
                if np.any(data == -3.0):
                    classes.append(constants.GRIB2_CLASSIFICATION[1])
                # some old files contain -999 as nodata value
                if np.any(data == -999.0):
                    classes.append(constants.GRIB2_CLASSIFICATION[2])
            elif np.any(data == -1.0):
                classes.append(constants.COG_CLASSIFICATION)

            if len(classes) > 0:
                band["classification:classes"] = classes
                # Add this if it gets accepted in v1.2:
                # see https://github.com/stac-extensions/classification/pull/34
                # band["classification:incomplete"] = True
            if len(classes) == 1:
                band["nodata"] = band["classification:classes"][0]["value"]

        proj_attrs = ProjectionExtension.ext(asset, add_if_missing=False)
        if shape:
            proj_attrs.shape = shape

        if transform:
            proj_attrs.transform = transform

        if epsg > 0 and not nogrib and not nocog:
            if isinstance(crs, int):
                proj_attrs.epsg = crs
            else:
                proj_attrs.epsg = None
                proj_attrs.projjson = crs

        asset.extra_fields["raster:bands"] = [band]

        return asset

    if basics.gzip:
        asset_href = cog.decompress(asset_href)

    if not nocog:
        epsg_string = "epsg:" + str(epsg) if epsg > 0 else None
        crs: Union[Dict[str, Any], int] = epsg if epsg > 0 else constants.PROJJSON
        cog_href = cog.convert(asset_href, reproject_to=epsg_string)

        band = create_band()

        asset = create_asset(
            cog_href,
            MediaType.COG,
            constants.COG_ROLES,
            band,
            crs,
            constants.ASSET_COG_TITLE,
        )
        item.add_asset(constants.ASSET_COG_KEY, asset)

    if not nogrib:
        band = create_band()

        asset = create_asset(
            asset_href,
            constants.GRIB2_MEDIATYPE,
            constants.GRIB2_ROLES,
            band,
            constants.PROJJSON,
            constants.ASSET_GRIB2_TITLE,
        )
        item.add_asset(constants.ASSET_GRIB2_KEY, asset)

    return item


def parse_filename(path: str) -> FileInfo:
    filename = os.path.basename(path)
    parts = constants.FILENAME_PATTERN.match(filename)
    if parts is None:
        raise ValueError(f"Filename is not valid: {filename}")

    year = int(parts.group(4))
    month = int(parts.group(5))
    day = int(parts.group(6))
    hour = int(parts.group(7))
    minute = int(parts.group(8))
    second = int(parts.group(9))
    time = datetime(year, month, day, hour, minute, second, tzinfo=timezone.utc)

    return FileInfo(
        id=parts.group(1),
        period=int(parts.group(2)),
        pass_no=int(parts.group(3)),
        datetime=time,
        gzip=False if parts.group(10) is None else True,
    )


def bbox_to_polygon(b: List[float]) -> Dict[str, Any]:
    return {
        "type": "Polygon",
        "coordinates": [
            [[b[0], b[3]], [b[2], b[3]], [b[2], b[1]], [b[0], b[1]], [b[0], b[3]]]
        ],
    }


def create_band() -> Dict[str, Any]:
    band: Dict[str, Any] = {}
    band["spatial_resolution"] = constants.RESOLUTION_M
    band["unit"] = constants.UNIT
    band["data_type"] = DataType.FLOAT64
    return band
